cardano-cli/src/Cardano/CLI/EraBased/HashChecking.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cardano.CLI.EraBased.HashChecking
+  ( checkCertificateHashes
+  , checkVotingProcedureHashes
+  , checkProposalHashes
+  )
+where
+
+import           Cardano.Api (Certificate (..), ExceptT, firstExceptT)
+import qualified Cardano.Api.Ledger as L
+import qualified Cardano.Api.Shelley as Shelley
+
+import           Cardano.CLI.Run.Hash (carryHashChecks)
+import           Cardano.CLI.Types.Common (MustCheckHash (..), PotentiallyCheckedAnchor (..))
+import           Cardano.CLI.Types.Errors.TxCmdError (TxCmdError (..))
+import qualified Cardano.Ledger.Api.Governance as L
+
+import           Control.Monad (forM_)
+import           Control.Monad.Trans.Except.Extra (left)
+
+-- | Check the hash of the anchor data against the hash in the anchor
+checkAnchorMetadataHash :: L.Anchor L.StandardCrypto -> ExceptT TxCmdError IO ()
+checkAnchorMetadataHash anchor =
+  firstExceptT (TxCmdHashCheckError $ L.anchorUrl anchor) $
+    carryHashChecks
+      ( PotentiallyCheckedAnchor
+          { pcaMustCheck = CheckHash
+          , pcaAnchor = anchor
+          }
+      )
+
+-- | Find references to anchor data and check the hashes are valid
+-- and they match the linked data.
+checkCertificateHashes :: Certificate era -> ExceptT TxCmdError IO ()
+checkCertificateHashes c =
+  case c of
```

Comment:
The code looks correct by itself, but this way of access of certificates is problematic. We should be using ledger patterns when accessing certificates. See how it's implemented here: https://github.com/IntersectMBO/cardano-api/blob/main/cardano-api/internal/Cardano/Api/Certificate.hs#L518

This function should be moved to `Cardano.Api.Certificate` in order to have handling of certificates in one place. We had multiple issues related to small mistakes related to ledger certificates, so considering future maintenance, it's better to keep it in one place.

Reply:
Addressed in: https://github.com/IntersectMBO/cardano-cli/commit/5784d620dc43415444d8c0f3b212c5c97c063ab6